File: 216-Combination Sum III-组合总和 III/216-Combination Sum III-组合总和 III.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
 let res = []
 let path = []
 
 var combinationSum3 = function(k, n) {
     res = []
     backtrace(k, n, 1)
     return res
 };
 
 var backtrace = function(k, n, index){
     if(path.length === k && sum(path) === n){
         res.push(Array.from(path))
         return
     }
 
     for(let i = index; i <= 9; i++){
         path.push(i)
         backtrace(k, n, i + 1)
         path.pop()
     }
 }
 
 var sum = function(arr){
     return arr.reduce(function(prev, curr, idx, arr){
     return prev + curr;
   });
 }

File: 216-Combination Sum III-组合总和 III/216-Combination Sum III-组合总和 III.java
class Solution {
    private List<List<Integer>> res;
    private int sum;
    private LinkedList<Integer> path = new LinkedList<Integer>();

    public List<List<Integer>> combinationSum3(int k, int n) {
        res = new ArrayList<>();
        sum = 0;
        backtrace(k, n, 1);
        return res;
    }

    public void backtrace(int k, int n, int index){
        if(sum > n){
            return;
        }
        if(path.size() == k){
            if(sum == n){
                res.add(new ArrayList<Integer>(){{addAll(path);}});
            }
            return;
        }

        for(int i = index; i <= 9 - (k - path.size()) + 1; i++){
            path.add(i);
            sum += i;
            backtrace(k, n, i + 1);
            path.removeLast();
            sum -= i;
        }
    } 
}